text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,335 @@
+
+[xargo]: https://github.com/japaric/xargo
+
+> What is the impact of not doing this?
+
+By not doing this, Rust will continue to be difficult to use for users and platforms "on the edge", such as new platform developers or embedded and WASM users.
+
+# Prior art
+[prior-art]: #prior-art
+
+* [RFC1133] - This RFC from 2015 proposed making cargo aware of std. I still need to review in more detail to find the parts and syntax that may solve some open questions.
+* [xargo] - This external tool was used to achieve a similar workflow as described above, limited to use with a nightly compiler
+* [Cargo Issue 5002] - This issue proposed a syntax for explicit dependency on std
+* [Cargo Issue 5003] - This issue discussed how to be backwards compatible with crates that don't explicitly mention std
+
+[RFC1133]: https://github.com/rust-lang/rfcs/pull/1133
+[Cargo Issue 5002]: https://github.com/rust-lang/cargo/issues/5002
+[Cargo Issue 5003]: https://github.com/rust-lang/cargo/issues/5003
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## How are dependencies (or non-dependency) on `core` and `std` specified?
+
+For example in a `no_core` or `no_std` crate, how would we tell Cargo **not** to build the `core` and/or `std` dependencies?
+
+## Should `std` be rebuilt if `core` is rebuilt?
+
+Is it necessary to rebuild `std` using the customized `core`, even if no changes to `std` are necessary?
+
+## Should Cargo obtain or verify the source code for `libcore` or `libstd`?
+
+Right now we depend on `rustup` to obtain the correct source code for these libraries, and we rely on the user not to tamper with the contents. Are these reasonable decisions?
+
+## Should the custom built `libcore` and `libstd` reside locally or globally?
+
+e.g., should the build artifacts be placed in `target/`, only usable by this project, or in `.cargo/`, to be possibly reused by multiple projects, if they happen to have the same settings?
+
+## How do we handle `libcore` and `libstd`'s `Cargo.lock` file?
+
+Right now these are built using the global lock file in `rust-lang/rust`. Should this always be true? How should Cargo handle this gracefully?
+
+## Should profile changes always prompt a rebuild of `core`/`std`?
+
+For example, if a user sets their debug build to use `opt-level = 'z'`, should this rebuild `core`/`std` to use that opt level? Or should an additional flag, such as `apply-to-sysroot` be required to opt-in to this behavior, unless otherwise needed?
+
+This could increase compile times for users that have set profile overrides, but have not previously needed a custom `core` or `std`.
+
+Another option in this area is to force the use of profile overrides, as specified by [RFC2822](https://github.com/rust-lang/rfcs/blob/master/text/2282-profile-dependencies.md).
+
+## Should providing a custom `core` or `std` require a nightly compiler?
```

Comment:
Yes, I think this should be considered a requirement. We cannot reasonably have stability and at the same time allow people to do with with a stable compiler.

Reply:
But isn't this already possible on stable? Just specify a `core`/`std` dependency in your `Cargo.toml` to override the default. Rust even automatically imports the `prelude` of the custom `std`.

We use this approach [in our `stm32f7-discorvery` crate](https://github.com/embed-rs/stm32f7-discovery/blob/e2bf713263791c028c2a897f2eb1830d7f09eceb/Cargo.toml#L21) to augment the `core` library with the [required `Future` implementations](augment) for using async/await on `no_std`.

Reply:
Ugh... @alexcrichton ^-- Was this intended and who intended it?

(the `Future` link is dead)

Reply:
Ah sorry, fixed it. Basically we provide our own implementation for the parts of [std::future](https://doc.rust-lang.org/std/future/index.html) that currently thread local storage (including the `await` macro).

Reply:
Also, part of the discussion on the pre-RFC was that overriding `std`/`core` on a stable compiler should be fine because actually implementing `std`/`core` will require using unstable features that will require a nightly compiler.

Reply:
Forcing unstable compiler just to be able to pick an std that works on embedded won't help regaining confidence that rust can be used in automotive

we need an alternative std that is stable.

Reply:
> EDIT: and if in the future it’s possible to provide an implementation of `std`/`core` without using any unstable features, why should the act of switching your dependencies to it require a nightly compiler?

When that happens and you don't need to set any lang items then that changes the equation entirely.

Reply:
> EDIT: and if in the future it’s possible to provide an implementation of `std`/`core` without using any unstable features, why should the act of switching your dependencies to it require a nightly compiler?

@Nemo157 I did mention that later on in the RFC, that I am of the opinion that IF you can develop your own `core` and/or `std` without "magic" features (which are required for things like `box` and `format!`), you should be able to use a stable compiler.

Reply:
@Centril What exactly are you worried about? `std` and `core` are just names that happen to usually be bound to libraries that use unstable features. If you want to override them with other features then you don't have to pay for those features. I can imagine that std-aware cargo is itself an unstable feature. But then the instability of overriding std/core is not a rustc concern, but a cargo concern: the instability is not the std and core are special and shouldn't be overridden, but that the way cargo adds implicit std/core dependencies is unstable.

In particular, I can imagine a situation where customary `std` and `core` still use unstable features, but it is possible to make a binary without using those unstable features. If the Cargo implicit stdlib deps mechanism is stable by then, then I think someone should be able to swap `std` and `core` out for stripped-down all stable code, even though building customized versions regular std still requires a nightly rustc.

Reply:
@Centril 

> > EDIT: and if in the future it’s possible to provide an implementation of `std`/`core` without using any unstable features, why should the act of switching your dependencies to it require a nightly compiler?
> 
> When that happens and you don't need to set any lang items then that changes the equation entirely.

Creating a stripped-down `std` library on top of `core` and `alloc` might be possible on stable Rust soon (when `alloc` is [stabilized](https://github.com/rust-lang/rfcs/pull/2480)). It would only work with `no_main` binaries and wouldn't contain all `std` features, but I don't think that it's problem that some crates would not build with it. In fact, it might be even preferable for environments like WASM, where e.g. threading is unavailable and accidentally using a crate that uses threads throws a _runtime_ error today.

Reply:
FWIW My thinking here is that Cargo simply won't do anything special in the long run when you're overriding core/std with a custom source. If it's compiling the sysroot std/core Cargo would enable unstable features manually for stable compilers (aka set the `RUSTC_BOOTSTRAP` env var). Otherwise if you crates can finagle some way to compile on stable but still override libstd, then more power to them! (aka you link to the real std and maybe wrap some of its functionality).

So in that sense I think Cargo, when compiling a custom core/std, just wouldn't do anything. If you use a bunch of unstable features that'll naturally require a nightly compiler, and if not then it's all stable anyway!

Reply:
> Otherwise if you crates can finagle some way to compile on stable but still override libstd, then more power to them! (aka you link to the real std and maybe wrap some of its functionality).

Should cargo provide a way to use the non-patched dependency from the patched one?

Reply:
> But isn't this already possible on stable?

@phil-opp Unlike `[patch.sysroot]`, having a crates.io dependency that happens to be named `core` or `std` only affects your own crate. And when doing so on stable, you’re very limited in what this dependency can do. In particular it cannot (re)define lang items.

> Forcing unstable compiler just to be able to pick an std that works on embedded

@aep Do you mean that works with async/await without thread-local storage? IMO this is a failure of async/await, not of the `core` v.s. `std` structure.

Reply:
@SimonSapin async is in core now, so we need to patch core.

Reply:
@aep Sorry, I haven’t followed all the recent development around async. Why do you need to patch `core`? Is there a path so that you don’t need to in the future?

In addition to being much preferable IMO, very practically for you, changes to async could happen sooner than stabilization of everything needed to define `core`.

Reply:
The only parts of `core` that the `async_await` feature uses are `core::{future::Future, task::{Poll, Context}}`, these are such trivial constructs that I don't see how you could change them while still having the `async_await` feature work? (I guess replacing `Context` would be possible since `async_await` only passes the type around, but doing that would definitely be regarded as operating outside standard procedures).

`async_await` also uses some functions from `std` currently, that is just an implementation defect (https://github.com/rust-lang/rust/issues/56974) and doesn't really seem like a good motivation for stabilizing the ability to provide your own `std` (it would be much better to put the energy towards fixing the defect so that you can just use the builtin async transform).